app/src/main/java/com/duckduckgo/app/browser/BrowserWebViewClient.kt

```diff
@@ -97,36 +96,28 @@ class BrowserWebViewClient(
 
     @UiThread
     override fun onPageStarted(webView: WebView, url: String?, favicon: Bitmap?) {
-        tempTimer = System.currentTimeMillis()
-        Timber.d("\nonPageStarted {\nurl: $url\nwebView.url: ${webView.url}\n}\n")
-        currentUrl = url
-
-        webViewClientListener?.let {
-            it.loadingStarted(url)
-            it.navigationOptionsChanged(WebViewNavigationOptions(webView.copyBackForwardList()))
+        webViewClientListener?.navigationStateChanged(WebViewNavigationState(webView.copyBackForwardList()))
+        if (url != null && url == lastPageStarted) {
+            webViewClientListener?.pageRefreshed(url)
         }
+        lastPageStarted = url
     }
 
     @UiThread
     override fun onPageFinished(webView: WebView, url: String?) {
-        currentUrl = url
-        webViewClientListener?.let {
-            it.loadingFinished(url)
-            it.navigationOptionsChanged(WebViewNavigationOptions(webView.copyBackForwardList()))
-        }
-
-        Timber.i("Page Load Time: ${System.currentTimeMillis() - tempTimer}ms for $url")
+        webViewClientListener?.navigationStateChanged(WebViewNavigationState(webView.copyBackForwardList()))
     }
 
     @WorkerThread
     override fun shouldInterceptRequest(webView: WebView, request: WebResourceRequest): WebResourceResponse? {
-        Timber.v("Intercepting resource ${request.url} on page $currentUrl")
-        return runBlocking { requestInterceptor.shouldIntercept(request, webView, currentUrl, webViewClientListener) }
+        val documentUrl = runBlocking { withContext(Dispatchers.Main) { webView.url } }
```

Comment:
The blocking call here is a bit icky, but a necessary evil as we can't use the currentUrl from onPageStarted. Happy to discuss alternatives.

Reply:
Why is there two `runBlocking`? couldn't it all be done within the same `runBlocking` block? e.g., 👇

    @WorkerThread
    override fun shouldInterceptRequest(webView: WebView, request: WebResourceRequest): WebResourceResponse? {
        return runBlocking {
            
            val documentUrl = withContext(Dispatchers.Main) { 
                webView.url 
            }
            
            Timber.v("Intercepting resource ${request.url} on page $documentUrl")
            requestInterceptor.shouldIntercept(request, webView, documentUrl, webViewClientListener)
        }
    }